contracts/TwabRewards.sol

```diff
@@ -0,0 +1,422 @@
+ * @notice Contract to distribute rewards to depositors in a pool.
+ */
+contract TwabRewards is ITwabRewards {
+    using SafeERC20 for IERC20;
+
+    /* ============ Global Variables ============ */
+
+    /// @notice Settings of each promotion.
+    mapping(uint256 => Promotion) internal _promotions;
+
+    /// @notice Latest promotion id.
+    uint256 internal _latestPromotionId;
+
+    /// @notice Keeps track of claimed rewards per user.
+    /// @dev _claimedEpochs[promotionId][user] => claimedEpochs
+    /// @dev We pack epochs claimed by a user into a uint256. So we can't store more than 256 epochs.
+    mapping(uint256 => mapping(address => uint256)) internal _claimedEpochs;
+
+    /* ============ Events ============ */
+
+    /**
+        @notice Emmited when a promotion is created.
+        @param promotionId Id of the newly created promotion
+    */
+    event PromotionCreated(uint256 indexed promotionId);
+
+    /**
+        @notice Emmited when a promotion is cancelled.
+        @param promotionId Id of the promotion being cancelled
+        @param amount Amount of tokens transferred to the promotion creator
+    */
+    event PromotionCancelled(uint256 indexed promotionId, uint256 amount);
+
+    /**
+        @notice Emmited when a promotion is extended.
+        @param promotionId Id of the promotion being extended
+        @param amount Amount of tokens transferred to the recipient address
+        @param numberOfEpochs New number of epochs after extending the promotion
+    */
+    event PromotionExtended(uint256 indexed promotionId, uint256 amount, uint256 numberOfEpochs);
+
+    /**
+        @notice Emmited when rewards have been claimed.
+        @param promotionId Id of the promotion in which epoch rewards were claimed
+        @param epochIds Ids of the epochs being claimed
+        @param amount Amount of tokens transferred to the recipient address
+    */
+    event RewardsClaimed(uint256 indexed promotionId, uint256[] indexed epochIds, uint256 amount);
+
+    /* ============ Modifiers ============ */
+
+    /// @dev Ensure that the caller is the creator of the currently active promotion.
+    /// @param _promotionId Id of the promotion to check
+    modifier onlyPromotionCreator(uint256 _promotionId) {
+        require(
+            msg.sender == _getPromotion(_promotionId).creator,
+            "TwabRewards/only-promotion-creator"
+        );
+        _;
+    }
+
+    /* ============ External Functions ============ */
+
+    /// @inheritdoc ITwabRewards
+    function createPromotion(
+        address _ticket,
+        address _token,
+        uint216 _tokensPerEpoch,
+        uint32 _startTimestamp,
+        uint32 _epochDuration,
+        uint8 _numberOfEpochs
+    ) external override returns (uint256) {
+        _requireTicket(_ticket);
+
+        uint256 _nextPromotionId = _latestPromotionId + 1;
+        _latestPromotionId = _nextPromotionId;
+
+        IERC20(_token).safeTransferFrom(
```

Comment:
checks effects interactions dicates you put this on line 108. cant see why it would be harmful here but good practise.

Reply:
Done: https://github.com/pooltogether/v4-periphery/pull/8/commits/57384eed7816668284e23f36f7a27fe988d69bbd